homeassistant/components/camera/__init__.py

```diff
@@ -0,0 +1,1133 @@
+            logging.getLogger(__name__).warning(
+                "Call to camera service did not specify an \
+                 action in the service data")
+            return
+
+        state = service.data.get('state', None)
+        if state is None:
+            logging.getLogger(__name__).warning(
+                "Call to camera service did not specify a \
+                 state in the service data")
+            return
+
+        target_cameras = component.extract_from_service(service)
+
+        for camera in target_cameras:
+            if action == SWITCH_ACTION_MOTION:
+                if (state == STATE_ON and not
+                        camera.is_motion_detection_enabled):
+                    camera.enable_motion_detection()
+                elif (
+                        state == STATE_OFF and
+                        camera.is_motion_detection_enabled):
+                    camera.disable_motion_detection()
+
+            elif action == SWITCH_ACTION_RECORD:
+                if (state == STATE_ON and not
+                        camera.is_recording):
+                    camera.record_stream()
+                elif (
+                        state == STATE_OFF and
+                        camera.is_recording):
+                    camera.stop_recording()
+
+            elif action == SWITCH_ACTION_SNAPSHOT:
+                if (state == STATE_ON and not
+                        camera.is_taking_snapshot):
+                    camera.take_snapshot()
+
+            camera.update_ha_state(True)
+
+    hass.services.register(
+        DOMAIN,
+        SERVICE_CAMERA,
+        handle_motion_detection_service)
+
+    return True
+
+
+# pylint: disable=too-many-instance-attributes
+# pylint: disable=too-many-public-methods
+class Camera(Entity):
+    """
+    Base class for cameras.
+    This is quite a large class but the camera component encompasses a lot of
+    functionality.  It should take care of most of the heavy lifting and
+    plumbing associated with adding support for additional models of camera.
+    If you are adding support for a new camera your entity class should inherit
+    from this.
+    """
+
+    def __init__(self, hass, device_info):
+        self.hass = hass
+        self.device_info = device_info
+        self.base_url = device_info.get('base_url')
+        if not self.base_url.endswith('/'):
+            self.base_url = self.base_url + '/'
+        self.username = device_info.get('username')
+        self.password = device_info.get('password')
+        self.is_streaming = False
+        self._is_detecting_motion = False
+        self._last_motion_detected = datetime.datetime.now()
+        self.last_connected_address = None
+        self._still_image_url = device_info.get('still_image_url', 'image.jpg')
+        # these are the camera functions and capabilities initialised
+        # to defaults, these should be overridden in derived classes
+        self._is_motion_detection_supported = False
```

Comment:
I think we can skip these variables and only have them as property methods that have hardcoded values. Platforms that extend the Camera class can then override it. This is the same pattern as used by the Entity class.